File: app/src/main/java/me/harshithgoka/youtubedl/YoutubeDL/VideoInfo.java
package me.harshithgoka.youtubedl.YoutubeDL;

import java.sql.Timestamp;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

public class VideoInfo {
    public String video_id;
    public String thumbnail_url;
    public String title;
    public String author;
    public int length;
    public int view_count;
    public List<Format> formats;
    public Timestamp last_updated;

    public String js_code;
    public String func_name;


    VideoInfo(String id, String title, String length, String view_count, String author, String thumbnail_url, Timestamp last_updated, List<Format> formats, String js_code, String func_name) {
        this.video_id = id;
        this.title = title;
        this.author = author;
        try {
            this.view_count = Integer.parseInt(view_count);
        }
        catch (Exception e) {
            // do nothing
            this.view_count = -1;
        }

        try {
            this.length = Integer.parseInt(length);
        }
        catch (Exception e) {
            // do nothing
            this.length = -1;
        }

        this.thumbnail_url = thumbnail_url;
        this.formats = new ArrayList<>();
        this.formats.addAll(formats);

        this.last_updated = last_updated;
    }

    public String getUrl() {
        return String.format(Locale.UK, "https://www.youtube.com/watch?v=%s", video_id);
    }

    public boolean isEquivalent(VideoInfo v1) {
        return video_id.equals(v1.video_id);
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof VideoInfo) {
            return this.isEquivalent((VideoInfo) obj);
        }
        return false;
    }
}

File: app/src/main/java/me/harshithgoka/youtubedl/YoutubeDL/VideoInfo.kt
package me.harshithgoka.youtubedl.YoutubeDL

import java.sql.Timestamp
import java.util.ArrayList
import java.util.Locale

class VideoInfo (var video_id: String, var title: String, length: String, view_count: String, var author: String, var thumbnail_url: String, var last_updated: Timestamp, formats: List<Format>) {
    var length: Int = -1
    var view_count: Int = -1
    var formats: MutableList<Format>

    val url: String
        get() = String.format(Locale.UK, "https://www.youtube.com/watch?v=%s", video_id)

    init {
        catchAll {
            this.view_count = Integer.parseInt(view_count)
        }

        catchAll {
            this.length = Integer.parseInt(length)
        }

        this.formats = ArrayList<Format>()
        this.formats.addAll(formats)
    }

    fun isEquivalent(v1: VideoInfo): Boolean {
        return video_id == v1.video_id
    }

    override fun equals(other: Any?): Boolean {
        return if (other is VideoInfo) {
            this.isEquivalent(other)
        } else false
    }

    private inline fun catchAll (action: () -> Unit) {
        try {
            action()
        }
        catch (e: Exception) {
            // do nothing
        }
    }

    override fun hashCode(): Int {
        return video_id.hashCode()
    }
}
